cpp/test/sg/rf_test.cu

```diff
@@ -40,370 +101,357 @@ struct RfInputs {
-
-    predict(
-      handle, forest, inference_data_d, params.n_inference_rows, params.n_cols, predicted_labels);
-    // Predict and compare against known labels
-    RF_metrics tmp = score(handle, forest, labels, params.n_inference_rows, predicted_labels);
-
-    CUDA_CHECK(cudaStreamSynchronize(stream));
-    CUDA_CHECK(cudaStreamDestroy(stream));
-
-    mse = tmp.mean_squared_error;
+    if (params.max_depth <= 1) { return; }
+    // accuracy is not guaranteed to improve with bootstrapping
+    if (params.bootstrap) { return; }
+    raft::handle_t handle(params.n_streams);
+    RfTestParams alt_params = params;
+    alt_params.max_depth--;
+    RF_metrics alt_metrics;
+    ML::RandomForestMetaData<DataT, LabelT> alt_forest;
+    TrainScore(handle,
+               alt_params,
+               X.data().get(),
+               X_transpose.data().get(),
+               y.data().get(),
+               &alt_forest,
+               alt_metrics);
+    double eps = 1e-8;
+    if (params.split_criterion == MSE) {
+      EXPECT_LE(training_metrics.mean_squared_error, alt_metrics.mean_squared_error + eps);
+    } else if (params.split_criterion == MAE) {
+      EXPECT_LE(training_metrics.mean_abs_error, alt_metrics.mean_abs_error + eps);
+    } else {
+      EXPECT_GE(training_metrics.accuracy, alt_metrics.accuracy);
+    }
   }
-
-  void SetUp() override { basicTest(); }
-
-  void TearDown() override
+  // Regularisation parameters are working correctly
+  void TestTreeSize()
   {
-    mse = -1.0f;  // reset mse
-    inference_data_h.clear();
-    labels_h.clear();
-
-    CUDA_CHECK(cudaFree(labels));
-    CUDA_CHECK(cudaFree(predicted_labels));
-    CUDA_CHECK(cudaFree(data));
-    CUDA_CHECK(cudaFree(inference_data_d));
-    delete forest;
+    for (int i = 0u; i < forest->rf_params.n_trees; i++) {
+      EXPECT_LE(forest->trees[i].depth_counter, params.max_depth);
+      if (params.max_leaves > 0) { EXPECT_LE(forest->trees[i].leaf_counter, params.max_leaves); }
+      EXPECT_LE(forest->trees[i].leaf_counter,
+                raft::ceildiv(params.n_rows, params.min_samples_leaf));
+    }
+  }
+  void TestDeterminism()
+  {
+    // Regression models use floating point atomics, so are not bitwise reproducible
+    bool is_regression = params.split_criterion == MSE || params.split_criterion == MAE;
+    if (is_regression) return;
+
+    // Repeat training
+    RF_metrics metrics;
+    raft::handle_t handle(params.n_streams);
+    ML::RandomForestMetaData<DataT, LabelT> alt_forest;
+    TrainScore(handle,
+               params,
+               X.data().get(),
+               X_transpose.data().get(),
+               y.data().get(),
+               &alt_forest,
+               metrics);
+
+    for (int i = 0u; i < forest->rf_params.n_trees; i++) {
+      EXPECT_EQ(forest->trees[i].sparsetree, alt_forest.trees[i].sparsetree);
+    }
+  }
+  void Test()
+  {
+    TestAccuracyImprovement();
+    // Bugs
+    // TestDeterminism();
```

Comment:
Is classification reproducibility still a problem?

Reply:
Yes. The node queue using atomic needs to be reworked in a later pr I think.